libraries/plugins/persistent_proposals/persistent_proposals_api.cpp

```diff
@@ -0,0 +1,189 @@
+    {
+        return get_proposed_transactions_impl(start, limit,
+                                              [&](const proposal_object& proposal) {
+                                                  return does_proposal_belong_to_account(proposal, account_id);
+                                              });
+    }
+
+    vector<proposal_update_object> get_proposal_updates( object_id_type proposal_id, object_id_type start, unsigned limit )const
+    {
+        FC_ASSERT(limit <= 100, "You may request at most 100 proposal updates at a time.");
+
+        vector<proposal_update_object> proposal_updates;
+
+        auto& proposal_updates_by_proposal_id = _database->get_index_type<proposal_update_index>().indices().get<by_proposal_id>();
+        proposal_update_multi_index_container::index<by_proposal_id>::type::iterator iter;
+        if (start == object_id_type())
+        {
+            iter = proposal_updates_by_proposal_id.lower_bound(std::make_tuple(proposal_id));
+        }
+        else
+        {
+            iter = proposal_updates_by_proposal_id.lower_bound(std::make_tuple(proposal_id, start));
+        }
+
+
+        while (iter != proposal_updates_by_proposal_id.end() &&
+               iter->proposal == proposal_id &&
+               proposal_updates.size() < limit)
+        {
+            proposal_updates.push_back(*iter);
+
+            ++iter;
+        }
+
+        return proposal_updates;
+    }
+
+private:
+    template <class Predicate>
+    vector<proposal_object> get_proposed_transactions_impl( object_id_type start, unsigned int limit, const Predicate& can_proposal_be_saved )const
+    {
+        FC_ASSERT(limit <= 100, "You may request at most 100 proposals at a time.");
+
+        const auto& proposals_by_id = _database->get_index_type<proposal_index>().indices().get<by_id>();
+        vector<proposal_object> proposals = get_proposals<proposal_object>(proposals_by_id, start, limit, can_proposal_be_saved);
+
+        const auto& persistent_proposals_by_original_id = _database->get_index_type<persistent_proposal_index>().indices().get<by_original_id>();
+        vector<persistent_proposal_object> expired_proposals = get_proposals<persistent_proposal_object>(persistent_proposals_by_original_id, start, limit, can_proposal_be_saved);
+        map_original_ids_to_proposal_id(&expired_proposals);
+
+        proposals.insert(proposals.end(), expired_proposals.begin(), expired_proposals.end());
+
+        std::sort(proposals.begin(), proposals.end(), [](const proposal_object& left, const proposal_object& right) {
+            return left.id < right.id;
+        });
+
+        if (proposals.size() > limit)
+        {
+            proposals.resize(limit);
+        }
+
+        return proposals;
+    }
+
+    static bool does_proposal_belong_to_account( const proposal_object& proposal, account_id_type account_id )
```

Comment:
For better performance, I think it's better to add another object or even more objects to store the required approvals and approved ones independently, actually this applies to the core as well.

Reply:
So you would rather setup a new object with new indeces rather than "filtering" through them when doing the call?
Basic question is: memory usage vs. computation, right?